core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala

```diff
@@ -473,6 +473,319 @@ class DAGSchedulerSuite
+    val reduceRdd = new MyRDD(sc, parts, List(shuffleDep))
+    submit(reduceRdd, (0 until parts).toArray)
+
+    val stage0Attempt = taskSets.last
+
+    // Make each task in stage 0 success, then fail all of stage 1
+    val completions = stage0Attempt.tasks.zipWithIndex.map{ case (task, idx) =>
+      (Success, makeMapStatus("host" + ('A' + idx).toChar, parts))
+    }.toSeq
+
+    complete(stage0Attempt, completions)
+
+    val stage1Attempt = taskSets.last
+    val failures = stage0Attempt.tasks.zipWithIndex.map{ case (task, idx) =>
+      (FetchFailed(makeBlockManagerId("hostA"), shuffleId, 0, idx, "ignored"), null)
+    }.toSeq
+
+    // Run Stage 1, this time with all fetchs failing
+    complete(stage1Attempt, failures)
+
+    // Resubmit and confirm that now all is well
+    scheduler.resubmitFailedStages()
+
+    assert(scheduler.runningStages.nonEmpty)
+    assert(!ended)
+
+    // Confirm job finished succesfully
+    val stage0Attempt2 = taskSets.last
+    val completions2 = stage0Attempt2.tasks.zipWithIndex.map{ case (task, idx) =>
+      (Success, makeMapStatus("host" + ('A' + idx).toChar, parts))
+    }.toSeq
+
+    complete(taskSets.last, completions2)
+
+    val stage1Attempt2 = taskSets.last
+    val completions3 = stage1Attempt2.tasks.zipWithIndex.map{ case (task, idx) =>
+      (Success, makeMapStatus("host" + ('A' + idx).toChar, parts))
+    }.toSeq
+
+    complete(taskSets.last, completions3)
+    sc.listenerBus.waitUntilEmpty(1000)
+    assert(!jobResult.isInstanceOf[JobFailed])
+    assert(ended === true)
+  }
+
+  /**
+   * In this test we demonstrate that only consecutive failures trigger a stage abort. In short, a
+   * job may fail multiple times, succeed once in a previously failed stage, then fail again, then
+   * complete both stages successfully without aborting.
+   */
+  test("Abort should only trigger after consecutive stage failures") {
+    // Create a new Listener to confirm that the listenerBus sees the JobEnd message
+    // when we abort the stage. This message will also be consumed by the EventLoggingListener
+    // so this will propagate up to the user.
+    var ended = false
+    var jobResult : JobResult = null
+    class EndListener extends SparkListener {
+      override def onJobEnd(jobEnd: SparkListenerJobEnd): Unit = {
+        jobResult = jobEnd.jobResult
+        ended = true
+      }
+    }
+
+    sc.listenerBus.addListener(new EndListener())
+
+    val shuffleOneRdd = new MyRDD(sc, 2, Nil).cache()
+    val shuffleDepOne = new ShuffleDependency(shuffleOneRdd, null)
+    val shuffleTwoRdd = new MyRDD(sc, 2, List(shuffleDepOne)).cache()
+    val shuffleDepTwo = new ShuffleDependency(shuffleTwoRdd, null)
+    val finalRdd = new MyRDD(sc, 1, List(shuffleDepTwo))
+    submit(finalRdd, Array(0))
+
+    // First, fail stage 0 multiple times after suceeding stage 2 and 1
+    for (attempt <- 1 to Stage.MAX_STAGE_FAILURES-1) {
+      // complete stage 2
+      complete(taskSets.last, Seq(
```

Comment:
so that example I linked to earlier has mistakes in its comments -- the first stage to run is stage 0, then stage 1, then stage 2.  Can you change these comments, and also update [these comments](https://github.com/apache/spark/blob/master/core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala#L813) while you're at it to avoid future confusion?